opensquirrel/default_gates.py

```diff
@@ -1,117 +1,211 @@
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        self.theta = Float(self.parameter).value
+        return BlochSphereRotation(qubit=q, axis=(0, 1, 0), angle=self.theta, phase=0)
 
 
 @named_gate
-def Rz(q: QubitLike, theta: Float) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=theta.value, phase=0)
+class Rz(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        self.theta = Float(self.parameter).value
+        return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=self.theta, phase=0)
 
 
 @named_gate
-def CNOT(control: QubitLike, target: QubitLike) -> ControlledGate:
-    return ControlledGate(control, X(target))
+class CNOT(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, control: QubitLike, target: QubitLike) -> ControlledGate:
+        return ControlledGate(control, X()(target))
 
 
 @named_gate
-def CZ(control: QubitLike, target: QubitLike) -> ControlledGate:
-    return ControlledGate(control, Z(target))
+class CZ(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, control: QubitLike, target: QubitLike) -> ControlledGate:
+        return ControlledGate(control, Z()(target))
 
 
 @named_gate
-def CR(control: QubitLike, target: QubitLike, theta: Float) -> ControlledGate:
-    return ControlledGate(
-        control,
-        BlochSphereRotation(qubit=target, axis=(0, 0, 1), angle=theta.value, phase=theta.value / 2),
-    )
+class CR(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, control: QubitLike, target: QubitLike) -> ControlledGate:
+        self.theta = Float(self.parameter).value
+        return ControlledGate(
+            control,
+            BlochSphereRotation(qubit=target, axis=(0, 0, 1), angle=self.theta, phase=self.theta / 2),
+        )
 
 
 @named_gate
-def CRk(control: QubitLike, target: QubitLike, k: SupportsInt) -> ControlledGate:
-    theta = 2 * math.pi / (2 ** Int(k).value)
-    return ControlledGate(control, BlochSphereRotation(qubit=target, axis=(0, 0, 1), angle=theta, phase=theta / 2))
+class CRk(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, control: QubitLike, target: QubitLike) -> ControlledGate:
+        self.theta = 2 * math.pi / (2 ** Int(self.parameter).value)
+        return ControlledGate(
+            control,
+            BlochSphereRotation(qubit=target, axis=(0, 0, 1), angle=self.theta, phase=self.theta / 2)
+        )
 
 
-default_bloch_sphere_rotations_without_params: list[Callable[[QubitLike], BlochSphereRotation]]
+default_bloch_sphere_rotations_without_params: list[type[NamedGateFunctor]]
```

Comment:
Would this not also include multi-qubit gates?

Reply:
Good point. We may still have `BlochSphereRotation`s as subclasses of `NamedGateFunctor`s.